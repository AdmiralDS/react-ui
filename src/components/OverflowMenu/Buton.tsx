import type { ButtonHTMLAttributes, ReactNode } from 'react';
import * as React from 'react';
import styled, { css } from 'styled-components';

import { ReactComponent as ButtonL } from '#src/components/OverflowMenu/svgs/ButtonL.svg';
import { ReactComponent as ButtonM } from '#src/components/OverflowMenu/svgs/ButtonM.svg';
import { ReactComponent as ButtonS } from '#src/components/OverflowMenu/svgs/ButtonS.svg';

export type Dimension = 'l' | 'm' | 's';

export const SIZE = {
  l: 36,
  m: 32,
  s: 24,
};

const focusStyle = css`
  border-radius: 50%;
  background-color: ${({ theme }) => theme.color['Opacity/Focus']};
`;

const ButtonComponent = styled.button<{ dimension: Dimension; menuOpened: boolean }>`
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  position: relative;
  cursor: pointer;
  border: 0;
  outline: none;
  padding: 0;
  background: transparent;
  -webkit-tap-highlight-color: transparent;
  overflow: visible;
  height: ${({ dimension }) => SIZE[dimension]}px;
  width: ${({ dimension }) => SIZE[dimension]}px;

  & svg {
    flex-shrink: 0;
  }

  & *[fill^='#'] {
    fill: ${({ theme }) => theme.color['Neutral/Neutral 50']};
  }

  &:hover {
    border-radius: 50%;
    background-color: ${({ theme }) => theme.color['Opacity/Hover']};
  }

  ${({ menuOpened }) => menuOpened && focusStyle}

  &:focus-visible {
    &:before {
      content: '';
      position: absolute;
      top: -4px;
      left: -4px;
      bottom: -4px;
      right: -4px;
      border: 2px solid ${({ theme }) => theme.color['Primary/Primary 60 Main']};
      border-radius: 4px;
    }
  }

  &:active {
    border-radius: 50%;
    background-color: ${({ theme }) => theme.color['Opacity/Focus']};
  }

  &:disabled {
    cursor: default;
    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color['Neutral/Neutral 30']};
    }
  }
`;

export const ButtonContent = styled.span`
  position: relative;
  display: inline-flex;
  align-items: center;
  height: 100%;
`;

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /** Элементы содержимого */
  children?: ReactNode;
  /** Размер кнопки */
  dimension?: Dimension;
  /** Состояние меню */
  menuOpened: boolean;
}

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ dimension = 'l', menuOpened, type = 'button', ...props }, ref) => {
    return (
      <ButtonComponent ref={ref} dimension={dimension} menuOpened={menuOpened} type={type} {...props}>
        <ButtonContent>
          {dimension === 'l' ? (
            <ButtonL width={24} height={24} aria-hidden />
          ) : dimension === 'm' ? (
            <ButtonM width={20} height={20} aria-hidden />
          ) : (
            <ButtonS width={16} height={16} aria-hidden />
          )}
        </ButtonContent>
      </ButtonComponent>
    );
  },
);
