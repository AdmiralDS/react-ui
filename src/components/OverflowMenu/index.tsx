import * as React from 'react';

import { ReactComponent as MoreHorizontalOutline } from '@admiral-ds/icons/build/system/MoreHorizontalOutline.svg';
import { ReactComponent as MoreVerticalOutline } from '@admiral-ds/icons/build/system/MoreVerticalOutline.svg';
import { passDropdownDataAttributes } from '#src/components/common/utils/splitDataAttributes';
import type { MenuModelItemProps } from '#src/components/Menu/MenuItem';
import type { DropMenuComponentProps, DropMenuStyleProps, RenderContentProps } from '#src/components/DropMenu';
import { DropMenu } from '#src/components/DropMenu';
import type { IconPlacementDimension } from '#src/components/IconPlacement';
import { IconPlacement } from '#src/components/IconPlacement';
import { refSetter } from '../common/utils/refSetter';

export type OverflowMenuDimension = 'l' | 'm' | 's';

export interface OverflowMenuProps
  extends Omit<React.HTMLAttributes<HTMLButtonElement>, 'onChange'>,
    Omit<DropMenuComponentProps, 'targetElement'>,
    DropMenuStyleProps {
  /** Выбранная опция */
  selected?: string;
  /**
   * @deprecated Помечено как deprecated в версии 3.2.0, будет удалено в 10.x.x версии.
   * Взамен используйте onSelectItem
   *
   * Колбек на изменение выбранной опции
   **/
  onChange?: (id: string) => void;
  /**
   * @deprecated Помечено как deprecated в версии 3.2.0, будет удалено в 10.x.x версии.
   * Взамен используйте onVisibilityChange
   *
   * Колбек на открытие меню
   **/
  onOpen?: () => void;
  /**
   * @deprecated Помечено как deprecated в версии 3.2.0, будет удалено в 10.x.x версии.
   * Взамен используйте nVisibilityChange
   *
   * Колбек на закрытие меню
   **/
  onClose?: () => void;
  /** Размер компонента */
  dimension?: OverflowMenuDimension;
  /** Отключение компонента */
  disabled?: boolean;
  /** Ориентация компонента */
  isVertical?: boolean;
  /** Опции выпадающего списка */
  items: Array<MenuModelItemProps>;
}

export const OverflowMenu = React.forwardRef<HTMLButtonElement, OverflowMenuProps>(
  (
    {
      dimension = 'l',
      disabled = false,
      alignSelf = 'flex-end',
      isVertical = false,
      disableSelectedOptionHighlight,
      selected,
      onSelectItem,
      active,
      onActivateItem,
      isVisible,
      onVisibilityChange,
      onClickOutside,
      onChange,
      onOpen,
      onClose,
      items,
      className = '',
      menuWidth,
      menuMaxHeight,
      dropContainerCssMixin,
      dropContainerClassName,
      dropContainerStyle,
      onForwardCycleApprove,
      onBackwardCycleApprove,
      renderTopPanel,
      renderBottomPanel,
      ...props
    },
    ref,
  ) => {
    const dropMenuProps = {
      ...passDropdownDataAttributes(props),
      renderTopPanel,
      renderBottomPanel,
      items,
      onChange,
      onOpen,
      onClose,
      isVisible,
      onVisibilityChange,
      onClickOutside,
      disableSelectedOptionHighlight,
      selected,
      onSelectItem,
      active,
      onActivateItem,
      menuMaxHeight,
      menuWidth,
      dropContainerCssMixin,
      dropContainerClassName,
      dropContainerStyle,
      alignSelf,
      onForwardCycleApprove,
      onBackwardCycleApprove,
    };

    const iconPlacementDimension = (dimension?: OverflowMenuDimension): IconPlacementDimension => {
      switch (dimension) {
        case 'l':
          return 'lBig';
        case 'm':
          return 'mBig';
        case 's':
          return 's';
        default:
          return 'lBig';
      }
    };

    return (
      <>
        <DropMenu
          {...dropMenuProps}
          dimension={dimension}
          disabled={disabled}
          renderContentProp={({ buttonRef, handleKeyDown, handleClick, menuState, disabled }: RenderContentProps) => {
            return (
              <IconPlacement
                {...props}
                ref={refSetter(ref, buttonRef as React.Ref<HTMLButtonElement>)}
                dimension={iconPlacementDimension(dimension)}
                disabled={disabled}
                highlightFocus={menuState}
                onClick={handleClick}
                aria-expanded={menuState}
                aria-haspopup={menuState}
                onKeyDown={handleKeyDown}
                className={className + ' overflow-menu-button-with-dropdown'}
              >
                {isVertical ? <MoreVerticalOutline /> : <MoreHorizontalOutline />}
              </IconPlacement>
            );
          }}
        />
      </>
    );
  },
);

OverflowMenu.displayName = 'OverflowMenu';
