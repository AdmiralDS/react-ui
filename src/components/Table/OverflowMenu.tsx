import * as React from 'react';

import { OverflowMenuWrapper } from './style';

interface OverflowMenuProps {
  tableWidth: number;
  row: any;
}

export const OverflowMenu: React.FC<OverflowMenuProps> = ({ tableWidth, row }) => {
  const oveflowMenuRef = React.useRef<HTMLDivElement>(null);
  const handleMenuOpen = () => {
    if (oveflowMenuRef.current) oveflowMenuRef.current.dataset.opened = 'true';
  };
  const handleMenuClose = () => {
    if (oveflowMenuRef.current) oveflowMenuRef.current.dataset.opened = 'false';
  };

  return (
    <OverflowMenuWrapper ref={oveflowMenuRef} data-opened={false} $offset={tableWidth}>
      {row.overflowMenuRender?.(row, handleMenuOpen, handleMenuClose)}
    </OverflowMenuWrapper>
  );
};
