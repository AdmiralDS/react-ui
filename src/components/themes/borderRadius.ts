import type { Shape } from '#src/components/themes/common';
export const ALL_BORDER_RADIUS_VALUES = [
  'Border radius 0',
  'Border radius 2',
  'Border radius 4',
  'Border radius 6',
  'Border radius 8',
  'Border radius 10',
] as const;

export function smallGroupBorderRadius(shape: Shape): string {
  switch (shape.borderRadiusKind) {
    case 'Border radius 0':
      return '0';
    case 'Border radius 2':
      return '2px';
    case 'Border radius 4':
      return '4px';
    case 'Border radius 6':
      return '4px';
    case 'Border radius 8':
      return '4px';
    case 'Border radius 10':
      return '4px';
    default:
      return '4px';
  }
}

export function mediumGroupBorderRadius(shape: Shape): string {
  switch (shape.borderRadiusKind) {
    case 'Border radius 0':
      return '0';
    case 'Border radius 2':
      return '2px';
    case 'Border radius 4':
      return '4px';
    case 'Border radius 6':
      return '6px';
    case 'Border radius 8':
      return '8px';
    case 'Border radius 10':
      return '10px';
    default:
      return '4px';
  }
}

export function largeGroupBorderRadius(shape: Shape): string {
  switch (shape.borderRadiusKind) {
    case 'Border radius 0':
      return '0';
    case 'Border radius 2':
      return '4px';
    case 'Border radius 4':
      return '8px';
    case 'Border radius 6':
      return '12px';
    case 'Border radius 8':
      return '16px';
    case 'Border radius 10':
      return '20px';
    default:
      return '8px';
  }
}
