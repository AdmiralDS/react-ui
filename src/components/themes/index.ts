export * from './light';
export * from './dark';
export type { Theme } from './common';
