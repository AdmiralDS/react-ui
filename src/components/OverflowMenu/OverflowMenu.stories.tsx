import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';
import styled from 'styled-components';
import { ReactComponent as MinusCircleOutline } from '@admiral-ds/icons/build/service/MinusCircleOutline.svg';

import { OverflowMenu } from '#src/components/OverflowMenu';
import { DropDownItem } from '#src/components/DropDownItem';

const Separator = styled.div`
  height: 20px;
  width: 8px;
`;

const DisplayBlock = styled.div`
  display: flex;
  align-items: flex-start;
`;

export default {
  title: 'Admiral-2.1/OverflowMenu',
  decorators: [withDesign],
  component: OverflowMenu,
  parameters: {
    layout: 'centered',
    design: [
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=39%3A25623',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=39%3A25674',
      },
    ],
  },
  argTypes: {
    dimension: {
      options: ['l', 'm', 's'],
      control: { type: 'radio' },
    },
    disabled: {
      control: { type: 'boolean' },
    },
    isVertical: {
      control: { type: 'boolean' },
    },
  },
} as ComponentMeta<typeof OverflowMenu>;

const Template1: ComponentStory<typeof OverflowMenu> = (args) => {
  const items: Array<any> = [
    {
      id: '1',
      display: 'Option one',
    },
    {
      id: '2',
      display: 'Option two',
    },
    {
      id: '3',
      display: 'Option three',
    },
    {
      id: '4',
      display: 'Option four',
    },
    {
      id: '5',
      display: 'Option five',
    },
    {
      id: '6',
      display: 'Option six',
    },
    {
      id: '7',
      display: 'Option seven',
    },
  ];

  const [selected, setSelected] = React.useState<string | null>(null);
  return (
    <>
      <OverflowMenu
        {...args}
        selected={selected}
        onChange={(id) => {
          console.log(`onChange('${id}')`);
          setSelected(id);
        }}
        onOpen={() => console.log('open menu')}
        onClose={() => console.log('close menu')}
        aria-label="Overflow Menu component"
      >
        {items.map((item) => {
          return (
            <DropDownItem role="option" key={item.id} id={item.id}>
              {item.display}
            </DropDownItem>
          );
        })}
      </OverflowMenu>
    </>
  );
};

const Template2: ComponentStory<typeof OverflowMenu> = (args) => {
  const items: Array<any> = [
    {
      id: '1',
      display: 'Option one',
    },
    {
      id: '2',
      display: (
        <div style={{ display: 'flex', width: '115px', justifyContent: 'space-between', alignItems: 'center' }}>
          <MinusCircleOutline width={20} height={20} />
          Option two
        </div>
      ),
    },
    {
      id: '3',
      display: 'Option three',
      disabled: true,
    },
    {
      id: '4',
      display: 'Option four',
    },
  ];

  const [selected, setSelected] = React.useState<string | null>(null);
  return (
    <DisplayBlock>
      <OverflowMenu
        {...args}
        selected={selected}
        onChange={(id) => setSelected(id)}
        onOpen={() => console.log('open menu')}
        onClose={() => console.log('close menu')}
        aria-label="Overflow Menu component"
      >
        {items.map((item) => {
          return (
            <DropDownItem role="option" key={item.id} id={item.id} disabled={item.disabled}>
              {item.display}
            </DropDownItem>
          );
        })}
      </OverflowMenu>
      <Separator />
      <OverflowMenu
        {...args}
        selected={selected}
        onChange={(id) => setSelected(id)}
        onOpen={() => console.log('open menu')}
        onClose={() => console.log('close menu')}
        aria-label="Overflow Menu component"
        disabled
      >
        {items.map((item) => {
          return (
            <DropDownItem role="option" key={item.id} id={item.id} disabled={item.disabled}>
              {item.display}
            </DropDownItem>
          );
        })}
      </OverflowMenu>
    </DisplayBlock>
  );
};

const Template3: ComponentStory<typeof OverflowMenu> = (args) => {
  const items: Array<any> = [
    {
      id: '1',
      display: 'Option one',
    },
    {
      id: '2',
      display: (
        <div style={{ display: 'flex', width: '115px', justifyContent: 'space-between', alignItems: 'center' }}>
          <MinusCircleOutline width={20} height={20} />
          Option two
        </div>
      ),
    },
    {
      id: '3',
      display: 'Option three',
      disabled: true,
    },
    {
      id: '4',
      display: 'Option four',
    },
  ];

  const [selected, setSelected] = React.useState<string | null>(null);
  return (
    <>
      <DisplayBlock>
        <OverflowMenu
          {...args}
          selected={selected}
          onChange={(id) => setSelected(id)}
          onOpen={() => console.log('open menu')}
          onClose={() => console.log('close menu')}
          aria-label="Overflow Menu component L"
        >
          {items.map((item) => {
            return (
              <DropDownItem role="option" key={item.id} id={item.id} disabled={item.disabled}>
                {item.display}
              </DropDownItem>
            );
          })}
        </OverflowMenu>
        <Separator />
        <OverflowMenu
          {...args}
          dimension="m"
          selected={selected}
          onChange={(id) => setSelected(id)}
          onOpen={() => console.log('open menu')}
          onClose={() => console.log('close menu')}
          aria-label="Overflow Menu component M"
        >
          {items.map((item) => {
            return (
              <DropDownItem role="option" key={item.id} id={item.id} disabled={item.disabled}>
                {item.display}
              </DropDownItem>
            );
          })}
        </OverflowMenu>
        <Separator />
        <OverflowMenu
          {...args}
          dimension="s"
          selected={selected}
          onChange={(id) => setSelected(id)}
          onOpen={() => console.log('open menu')}
          onClose={() => console.log('close menu')}
          aria-label="Overflow Menu component S"
        >
          {items.map((item) => {
            return (
              <DropDownItem role="option" key={item.id} id={item.id} disabled={item.disabled}>
                {item.display}
              </DropDownItem>
            );
          })}
        </OverflowMenu>
      </DisplayBlock>
      <Separator />
      <DisplayBlock>
        <OverflowMenu
          {...args}
          isVertical
          selected={selected}
          onChange={(id) => {
            console.log(`onChange('${id}')`);
            setSelected(id);
          }}
          onOpen={() => console.log('open menu')}
          onClose={() => console.log('close menu')}
          aria-label="Overflow Menu component vertical L"
        >
          {items.map((item) => {
            return (
              <DropDownItem role="option" key={item.id} id={item.id}>
                {item.display}
              </DropDownItem>
            );
          })}
        </OverflowMenu>
        <Separator />
        <OverflowMenu
          {...args}
          dimension="m"
          isVertical
          selected={selected}
          onChange={(id) => {
            console.log(`onChange('${id}')`);
            setSelected(id);
          }}
          onOpen={() => console.log('open menu')}
          onClose={() => console.log('close menu')}
          aria-label="Overflow Menu component vertical M"
        >
          {items.map((item) => {
            return (
              <DropDownItem role="option" key={item.id} id={item.id}>
                {item.display}
              </DropDownItem>
            );
          })}
        </OverflowMenu>
        <Separator />
        <OverflowMenu
          {...args}
          dimension="s"
          isVertical
          selected={selected}
          onChange={(id) => {
            console.log(`onChange('${id}')`);
            setSelected(id);
          }}
          onOpen={() => console.log('open menu')}
          onClose={() => console.log('close menu')}
          aria-label="Overflow Menu component vertical S"
        >
          {items.map((item) => {
            return (
              <DropDownItem role="option" key={item.id} id={item.id}>
                {item.display}
              </DropDownItem>
            );
          })}
        </OverflowMenu>
      </DisplayBlock>
    </>
  );
};

export const OverflowMenuBase = Template1.bind({});
OverflowMenuBase.args = {};
OverflowMenuBase.storyName = 'OverflowMenu. Базовый пример.';
OverflowMenuBase.parameters = {
  docs: {
    source: {
      type: 'code',
    },
  },
};

export const OverflowMenuOptions = Template2.bind({});
OverflowMenuOptions.args = {};
OverflowMenuOptions.storyName = 'OverflowMenu. Задизейбленные и кастомизированные опции.';
OverflowMenuOptions.parameters = {
  docs: {
    source: {
      type: 'code',
    },
  },
};

export const OverflowMenuSizesOrientation = Template3.bind({});
OverflowMenuSizesOrientation.args = {};
OverflowMenuSizesOrientation.storyName = 'OverflowMenu. Размеры и ориентация.';
OverflowMenuSizesOrientation.parameters = {
  docs: {
    source: {
      type: 'code',
    },
  },
};
