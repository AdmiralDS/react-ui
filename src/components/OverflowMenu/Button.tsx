import type { ButtonHTMLAttributes, ReactNode } from 'react';
import * as React from 'react';
import styled, { css } from 'styled-components';

import { ReactComponent as MoreHorizontalOutline } from '@admiral-ds/icons/build/system/MoreHorizontalOutline.svg';
import { ReactComponent as MoreVerticalOutline } from '@admiral-ds/icons/build/system/MoreVerticalOutline.svg';
import { mediumGroupBorderRadius } from '#src/components/themes/borderRadius';

export type Dimension = 'l' | 'm' | 's';

export const SIZE = {
  l: 36,
  m: 32,
  s: 24,
};

const ICON_SIZE = {
  l: 24,
  m: 20,
  s: 16,
};

const focusStyle = css`
  border-radius: 50%;
  background-color: ${({ theme }) => theme.color['Opacity/Focus']};
`;

const ButtonComponent = styled.button<{ dimension: Dimension; menuOpened: boolean }>`
  display: flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  position: relative;
  cursor: pointer;
  border: 0;
  outline: none;
  padding: 0;
  background: transparent;
  -webkit-tap-highlight-color: transparent;
  overflow: visible;
  height: ${({ dimension }) => SIZE[dimension]}px;
  width: ${({ dimension }) => SIZE[dimension]}px;

  &:hover:not(:disabled) {
    border-radius: 50%;
    background-color: ${({ theme }) => theme.color['Opacity/Hover']};
  }

  ${({ menuOpened }) => menuOpened && focusStyle}

  &:focus-visible {
    &:before {
      content: '';
      position: absolute;
      top: -4px;
      left: -4px;
      bottom: -4px;
      right: -4px;
      border: 2px solid ${({ theme }) => theme.color['Primary/Primary 60 Main']};
      border-radius: ${(p) => mediumGroupBorderRadius(p.theme.shape)};
    }
  }

  &:active:not(:disabled) {
    border-radius: 50%;
    background-color: ${({ theme }) => theme.color['Opacity/Focus']};
  }

  &:disabled {
    cursor: default;
    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color['Neutral/Neutral 30']};
    }
  }
`;

export const ButtonContent = styled.span`
  position: relative;
  display: inline-flex;
  align-items: center;
  width: 100%;
  height: 100%;
`;

const IconContainer = styled.div<{ dimension: Dimension }>`
  height: ${({ dimension }) => ICON_SIZE[dimension]}px;
  width: ${({ dimension }) => ICON_SIZE[dimension]}px;
  margin: auto;

  & > svg {
    height: 100%;
    width: 100%;
    flex-shrink: 0;

    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color['Neutral/Neutral 50']};
    }
  }
`;

export const OverflowMenuIcon = React.forwardRef<HTMLDivElement, { dimension: Dimension; isVertical: boolean }>(
  ({ dimension = 'l', isVertical = false }, ref) => {
    const Icon = isVertical ? MoreVerticalOutline : MoreHorizontalOutline;

    return (
      <IconContainer ref={ref} dimension={dimension}>
        <Icon aria-hidden />
      </IconContainer>
    );
  },
);

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  /** Элементы содержимого */
  children?: ReactNode;
  /** Размер кнопки */
  dimension?: Dimension;
  /** Состояние меню */
  menuOpened: boolean;
}

export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ dimension = 'l', menuOpened, type = 'button', children, ...props }, ref) => {
    return (
      <ButtonComponent ref={ref} dimension={dimension} menuOpened={menuOpened} type={type} {...props}>
        <ButtonContent>{children}</ButtonContent>
      </ButtonComponent>
    );
  },
);
