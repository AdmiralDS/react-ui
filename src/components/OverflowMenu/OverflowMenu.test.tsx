import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'styled-components';

import { OverflowMenu } from '../OverflowMenu';
import { LIGHT_THEME } from '../themes';

describe('MenuButton', () => {
  jest.useFakeTimers();
  beforeEach(() => {
    jest.clearAllTimers();
  });

  const requiredProps = {
    selected: '1',
    onChange: (id: string) => {
      /**/
    },
    options: [
      { id: '1', display: 'one' },
      { id: '2', display: 'two' },
      { id: '3', display: 'three' },
    ],
  };
  const Component = (props: any) => (
    <ThemeProvider theme={LIGHT_THEME}>
      <OverflowMenu {...props} {...requiredProps} data-testid="btn" />
    </ThemeProvider>
  );

  it('should render component', () => {
    const wrapper = render(<Component />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should show menu when user clicks on button', () => {
    render(<Component />);
    act(() => {
      fireEvent.click(screen.getByTestId('btn'));
    });
    expect(screen.getAllByRole('listbox')).toHaveLength(1);
  });

  it('should show menu when user presses Enter', () => {
    render(<Component />);
    userEvent.tab();
    const btn = screen.getByTestId('btn');
    act(() => {
      fireEvent.keyDown(btn, { key: 'Enter', code: 'Enter' });
    });
    expect(screen.getAllByRole('listbox')).toHaveLength(1);
  });

  it('should show menu when user presses Space', () => {
    render(<Component />);
    userEvent.tab();
    const btn = screen.getByTestId('btn');
    act(() => {
      fireEvent.keyDown(btn, { key: ' ', code: 'Space' });
    });
    expect(screen.getAllByRole('listbox')).toHaveLength(1);
  });

  it('should show menu when user presses ArrowDown', () => {
    render(<Component />);
    userEvent.tab();
    const btn = screen.getByTestId('btn');
    act(() => {
      fireEvent.keyDown(btn, { key: 'ArrowDown', code: 'ArrowDown' });
    });
    expect(screen.getAllByRole('listbox')).toHaveLength(1);
  });

  it('should hide menu when user presses Escape', () => {
    render(<Component />);
    userEvent.tab();
    const btn = screen.getByTestId('btn');
    act(() => {
      fireEvent.keyDown(btn, { key: 'ArrowDown', code: 'ArrowDown' });
    });
    const menu = screen.getAllByRole('listbox')[0];
    act(() => {
      fireEvent.keyDown(menu, { key: 'Escape', code: 'Escape' });
    });
    expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
  });
});
