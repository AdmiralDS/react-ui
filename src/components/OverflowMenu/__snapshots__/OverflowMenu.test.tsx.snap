// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OverflowMenu should render component 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c3 {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
}

.c3>svg {
  height: 24px;
  width: 24px;
}

.c2 {
  width: 36px;
  height: 36px;
  border-radius: 50%;
  background-color: transparent;
  pointer-events: none;
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
}

.c0 {
  position: relative;
  padding: 0;
  margin: 6px;
  box-sizing: border-box;
  border: none;
  background-color: transparent;
  appearance: none;
  flex: 0 0 auto;
  height: 24px;
  width: 24px;
  border-radius: var(--admiral-border-radius-Small, 4px);
  overflow: visible;
  cursor: pointer;
}

.c0 >* {
  pointer-events: none;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0:disabled *[fill^='#'] {
  fill: var(--admiral-color-Neutral_Neutral30, #B2B5B8);
}

.c0:not(:disabled):focus-visible {
  outline-offset: 2px;
  outline: var(--admiral-color-Primary_Primary60Main, #0062FF) solid 2px;
}

.c0:not(:disabled):focus >.c1 {
  background-color: transparent;
}

.c0:not(:disabled):hover >.c1 {
  background-color: var(--admiral-color-Opacity_Hover, #0000000D);
}

.c0:not(:disabled):active >.c1 {
  background-color: var(--admiral-color-Opacity_Press, #0000001F);
}

.c0:not(:disabled):focus-visible >.c1 {
  background-color: transparent;
}

<body>
    <div>
      <button
        aria-expanded="false"
        aria-haspopup="false"
        class="c0  overflow-menu-button-with-dropdown"
        data-testid="btn"
        type="button"
      >
        <div
          aria-hidden="true"
          class="c1 c2"
        />
        <div
          aria-hidden="true"
          class="c3"
        >
          <div />
        </div>
      </button>
    </div>
  </body>,
  "container": .c3 {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
}

.c3>svg {
  height: 24px;
  width: 24px;
}

.c2 {
  width: 36px;
  height: 36px;
  border-radius: 50%;
  background-color: transparent;
  pointer-events: none;
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
}

.c0 {
  position: relative;
  padding: 0;
  margin: 6px;
  box-sizing: border-box;
  border: none;
  background-color: transparent;
  appearance: none;
  flex: 0 0 auto;
  height: 24px;
  width: 24px;
  border-radius: var(--admiral-border-radius-Small, 4px);
  overflow: visible;
  cursor: pointer;
}

.c0 >* {
  pointer-events: none;
}

.c0:disabled {
  cursor: not-allowed;
}

.c0:disabled *[fill^='#'] {
  fill: var(--admiral-color-Neutral_Neutral30, #B2B5B8);
}

.c0:not(:disabled):focus-visible {
  outline-offset: 2px;
  outline: var(--admiral-color-Primary_Primary60Main, #0062FF) solid 2px;
}

.c0:not(:disabled):focus >.c1 {
  background-color: transparent;
}

.c0:not(:disabled):hover >.c1 {
  background-color: var(--admiral-color-Opacity_Hover, #0000000D);
}

.c0:not(:disabled):active >.c1 {
  background-color: var(--admiral-color-Opacity_Press, #0000001F);
}

.c0:not(:disabled):focus-visible >.c1 {
  background-color: transparent;
}

<div>
    <button
      aria-expanded="false"
      aria-haspopup="false"
      class="c0  overflow-menu-button-with-dropdown"
      data-testid="btn"
      type="button"
    >
      <div
        aria-hidden="true"
        class="c1 c2"
      />
      <div
        aria-hidden="true"
        class="c3"
      >
        <div />
      </div>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
